vshard/router/init.lua

```diff
@@ -873,36 +888,158 @@ local function router_map_callrw(router, func, args, opts)
                             router.total_bucket_count - bucket_count)
         goto fail
     end
+    do return timeout, nil, nil, rid, futures, replicasets end
+
+    ::fail::
+    return nil, err, err_id, rid, futures
+end
+
+--
```

Comment:
They aren't super similar though. Yes, they all do *"send requests async, collect responses blocking"*. But they use the results quite differently.

We could extract this send+collect into a function, but it isn't going to be big, not much reuse. And it would need to construct a map `{replicaset_id : result}`, while right now we are avoiding that by handling each result right after getting it from the futures one by one.

Still, it is an interesting idea to try, I agree. Perhaps not in this PR though. Its size is already getting out of hand.